src/main/java/com/faforever/client/vault/replay/LiveReplayFilterController.java

```diff
@@ -0,0 +1,77 @@
+package com.faforever.client.vault.replay;
+
+import com.faforever.client.domain.GameBean;
+import com.faforever.client.fx.Controller;
+import com.faforever.client.fx.JavaFxUtil;
+import com.faforever.commons.lobby.GameStatus;
+import com.faforever.commons.lobby.GameType;
+import javafx.beans.InvalidationListener;
+import javafx.beans.WeakInvalidationListener;
+import javafx.beans.binding.BooleanBinding;
+import javafx.collections.transformation.FilteredList;
+import javafx.scene.control.CheckBox;
+import javafx.scene.control.ToggleButton;
+import javafx.scene.layout.VBox;
+import lombok.RequiredArgsConstructor;
+import org.springframework.beans.factory.config.ConfigurableBeanFactory;
+import org.springframework.context.annotation.Scope;
+import org.springframework.stereotype.Component;
+
+@Component
+@Scope(ConfigurableBeanFactory.SCOPE_PROTOTYPE)
+@RequiredArgsConstructor
+public class LiveReplayFilterController implements Controller<VBox> {
+
+  public VBox root;
+  public CheckBox customGamesCheckbox;
+  public CheckBox matchmakerCheckbox;
+  public CheckBox coopGamesCheckbox;
+  public CheckBox moddedGamesCheckbox;
+
+  private InvalidationListener selectedCheckboxListener;
+  private FilteredList<GameBean> gameList;
+  private ToggleButton filterButton;
+  private BooleanBinding filterApplied;
+
+  @Override
+  public void initialize() {
+    filterApplied = customGamesCheckbox.selectedProperty().not()
+        .or(matchmakerCheckbox.selectedProperty().not())
+        .or(coopGamesCheckbox.selectedProperty().not())
+        .or(moddedGamesCheckbox.selectedProperty().not());
+
+    selectedCheckboxListener = (observable) -> {
+      updateFilterButtonState();
+      gameList.setPredicate(game -> {
+        boolean playing = game.getStatus() == GameStatus.PLAYING;
+        boolean custom = customGamesCheckbox.isSelected() || game.getGameType() != GameType.CUSTOM;
+        boolean matchmaker = matchmakerCheckbox.isSelected() || game.getGameType() != GameType.MATCHMAKER;
+        boolean coopGame = coopGamesCheckbox.isSelected() || game.getGameType() != GameType.COOP;
+        boolean moddedGame = moddedGamesCheckbox.isSelected() || game.getSimMods().isEmpty();
+        return playing && custom && matchmaker && coopGame && moddedGame;
+      });
+    };
+  }
+
+  @Override
+  public VBox getRoot() {
+    return root;
+  }
+
+  public void setGameList(FilteredList<GameBean> gameList) {
+    this.gameList = gameList;
+    JavaFxUtil.addListener(customGamesCheckbox.selectedProperty(), new WeakInvalidationListener(selectedCheckboxListener));
+    JavaFxUtil.addListener(matchmakerCheckbox.selectedProperty(), new WeakInvalidationListener(selectedCheckboxListener));
+    JavaFxUtil.addListener(coopGamesCheckbox.selectedProperty(), new WeakInvalidationListener(selectedCheckboxListener));
+    JavaFxUtil.addAndTriggerListener(moddedGamesCheckbox.selectedProperty(), new WeakInvalidationListener(selectedCheckboxListener));
+  }
```

Comment:
Rather than this taking in a game list can we have it just have a getPredicate() method and then also the ability to listen to changes in the predicate so that the class using it still controls the list.

Also we can just name this class GameFilterController since we could also use this other places.

Reply:
I am having trouble with that. I can not figure out how to do it. Could you give a pseudocode?

Reply:
Could be something like this
```
class GameFilterController {
    private ObjectProperty<Predicate<GameBean>> predicateProperty;

    public void initalize() {
      predicateProperty.setValue(game -> {
        boolean playing = game.getStatus() == GameStatus.PLAYING;
        boolean custom = customGamesCheckbox.isSelected() || game.getGameType() != GameType.CUSTOM;
        boolean matchmaker = matchmakerCheckbox.isSelected() || game.getGameType() != GameType.MATCHMAKER;
        boolean coopGame = coopGamesCheckbox.isSelected() || game.getGameType() != GameType.COOP;
        boolean moddedGame = moddedGamesCheckbox.isSelected() || game.getSimMods().isEmpty();
        return playing && custom && matchmaker && coopGame && moddedGame;
      });

      public ObjectProperty<> getPredicateProperty() {
        return predicateProperty;
      }

      public Predicate<> getPredicate() {
        return predicateProperty.get();
      }
}

Where used

gameList.setPredicate(gameFilterController.getPredicate());
```

Here you don't even really need the listener beyond setting if the filter is actually being applied or not. But here you would just get the value of the predicate from the controller and could apply it to whatever game list. Could even use it in multiple places.

This would also be easy to extend for things like #1946. Also the predicateProperty is only needed if the predicate actually changes in this case nothing will change the predicate but could make extension easy.

Reply:
> Here you don't even really need the listener
Sorry for my stupidity.Then how do I tell the list to filter items when the checkbox changes its state without `predicate` listener?

Reply:
Wait...

Reply:
Yeah unless I misunderstand how the predicate works I think it should automatically pick up the change. If it doesn't then, you can add a filterStateChange property and listen to that to refresh the predicate

Reply:
> I think it should automatically pick up the change

Nope. It doesn't work like that

Reply:
> This would also be easy to extend for things like https://github.com/FAForever/downlords-faf-client/issues/1946. Also the predicateProperty is only needed if the predicate actually changes in this case nothing will change the predicate but could make extension easy.

OK. I will try to make the class extensible.

Reply:
> > I think it should automatically pick up the change
> 
> Nope. It doesn't work like that

Hmm unfortunate but in that case just add a predicate changed listener or just update the predicate property value and attach a listener to that and you can detect when the predicate is changed